### DISCUSSION TITLE
爬蟲限流

### DISCUSSION THREAD

**Discussion by lazywulf** [1101-GS4719-C/Python-Community-GS4719-C#1735] [General]
正常情況下，使用爬蟲時，如果爬太快 (一秒發上萬個request)，目標伺服器應該會阻止該爬蟲的訪問。你的終端機應該會被403噴好噴滿。但......我們當然想要越快越好，大概會想要作異步爬蟲 (或用多線程搞一搞)。那這時怎麼限制爬蟲速度呢？

本人的經驗：使用 semaphore 限流。

在解釋這之前，我們來複習一下race condition。Race condition 用來描述一個系統或程序的輸出依賴於不受控制的事件出現順序或者出現時機。此詞源自於兩個訊號試著彼此競爭，來影響誰先輸出。舉個例子，如果今天有兩個人要吃義大利麵，桌上有一隻叉子，他們倆勢必得搶叉子。這裡的兩人就是程序們，叉子是資源。

這裡，一次能發出 request 的量便是資源了。只要我們能限制發出的request量，並且在「資源」釋出當下發下一個request，我們便可以實現高速且不被伺服器擋掉的爬蟲。

Semaphore 是一個同步物件，用於保持在 0 至指定最大值之間的一個計數值。當執行緒完成一次對該 semaphore 物件的等待 (wait) 時，該計數值減一；當執行緒完成一次對 semaphore 物件的釋放 (release) 時，計數值加一。Semaphore 物件的計數值大於 0，為 signaled 狀態；計數值等於 0，為 nonsignaled 狀態。也就是說，當計數值為 0 時，處於 nonsignald 狀態的 semaphore 是無法供執行緒使用的，除非該 semaphore 再次轉為 signaled 狀態。

我們接下來只要實驗一下對方伺服器的訪問上限，再依照該上限定 semaphore 的最大值，就完成了喔~

下面是參考用的東東：
[Synchronization](https://hackmd.io/@combo-tw/Linux-%E8%AE%80%E6%9B%B8%E6%9C%83/%2F%40combo-tw%2FB16v1IRBB#Semaphore)
[Race Condition](https://zh.wikipedia.org/zh-tw/%E7%AB%B6%E7%88%AD%E5%8D%B1%E5%AE%B3)
[Semaphore](https://zh.wikipedia.org/wiki/%E4%BF%A1%E5%8F%B7%E9%87%8F)

**Comment by Chengbosco** [comment]
原來有這種方法，害我想說如果要抓股市的資料10秒一次也太慢了WWW。
所以就像兩隻叉子一個撈完另一個馬上撈?

**Reply by lazywulf** [reply]
沒錯喔！叉子數量就是 semaphore 的最大值喔！撈完換下一位吃，不要叫大家慢慢等。
